### ISSUE TITLE
Record property nullability check uses base class rather than child class's property

### ISSUE THREAD

**Issue by SolalPirelli** [dotnet/roslyn#79864]
**Version Used**: VS Community 17.14.11+36401.2 (C# Tools 4.14.0-3.25359.3+6dbcfd2f553ef1ef0ac48878ab17d18c1360735d)

**Steps to Reproduce**:

```csharp
record Base(string? Property);

record Child(string Property) : Base(Property)
{
    public override string ToString() => Property;
}
```

**Diagnostic Id**: CS8603: 'Property' may be null here"

**Expected Behavior**: No warnings since `Child.Property` cannot be null

**Actual Behavior**: Warning in the use of `Property` in the `ToString()` body, with tooltip stating `Base.Property` (rather than the derived one) can be null:

<img width="451" height="208" alt="Image" src="https://github.com/user-attachments/assets/37ab7c58-f4c6-41fb-b1d8-1c68e5798c7b" />

**Comment by Youssef1313** [comment]
I think this is no different from this case:

```csharp
class Base
{
    public Base(string? property)
       => Property = property;

    public string? Property { get; }
}

class Child : Base
{
    public Child(string property)
        : base(property)
    {
    }

    public override string ToString()
        => Property;
}
```

which also produces a warning.

**Comment by SolalPirelli** [comment]
In terms of desugared code, sure, but with records the compiler should be able to avoid the warning since it can see where `Property` is defined, no?

**Comment by RikkiGibson** [comment]
The language doesn't verify that your base clause is passing thru the parameter value as-is and doesn't verify that the base property always returns the parameter value. So it's not able to assume that a `Property` accessed on `Child` is non-null.

For example, you are allowed to do this:
```cs
record Base(string? Property);

record Child(string Property) : Base(Verify(Property))
{
    private static string? Verify(string s) => s is "bad" ? null : s;
}
```

Or this:
```cs
record Base(string? Property)
{
    public string Property { get; init; } = Property is "bad" ? null : Property;
}

record Child(string Property) : Base(Property)
{
}
```

**Comment by RikkiGibson** [comment]
Closing this out as by-design. Feel free to follow up with any further questions or concerns.

**Comment by SolalPirelli** [comment]
I thought about it more and even if the compiler detected the lack of custom Property and other shenanigans, there's a simpler way to bypass it:

```
Child child = F(new Child("")); // a Child with a null Property!

static T F<T>(T t) where T : Base
{
    return t with { Property = null };
}

record Base(string? Property);

record Child(string Property) : Base(Property);
```

Too bad :/ I'll have to live with a few `!` in my codebase.
